File: Data Structures & Algorithms/combinations-of-a-phone-number/submission-1.cpp
class Solution {
    vector<string> output;
    vector<string> dict_map = {
        "",    "",    "abc",  "def",  "ghi",
        "jkl", "mno", "pqrs", "tuv",  "wxyz"
    };

    void dfs(int idx, string &res, const string &digits){
        if (idx == (int)digits.size()){
            output.push_back(res);
            return;
        }
        int d = digits[idx] - '0';
        const string &letters = dict_map[d];
        for (char ch: letters){
            res.push_back(ch);
            dfs(idx + 1, res, digits);
            res.pop_back();
        }
    }
public:
    vector<string> letterCombinations(string digits) {
        output.clear();
        if (digits.empty()) return output;

        string res;
        dfs(0, res, digits);
        return output;
    }
};

File: Data Structures & Algorithms/count-paths/submission-1.py
class Solution:
    def uniquePaths(self, m: int, n: int) -> int:
        
        dp = {}
        def dfs(i, j):
            if i < 0 or i > m - 1 or j < 0 or j > n - 1:
                return 0

            if i == m - 1 and j == n - 1:
                return 1

            if (i, j) in dp:
                return dp[(i, j)]
            
            dp[(i, j)] = dfs(i + 1, j) + dfs(i, j + 1)
            return dp[(i, j)]

        
        return dfs(0, 0)